### PR TITLE
Add proxy authn

### PR DESCRIPTION
**Pull request by marcinsbd**

### CODE REVIEW

File: http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java [hunk 1]
```diff
@@ -301,7 +303,13 @@ protected void configure(SelectableChannel selectable)
         }
         HostAndPort httpProxy = config.getHttpProxy();
         if (httpProxy != null) {
-            httpClient.getProxyConfiguration().addProxy(new HttpProxy(new Address(httpProxy.getHost(), httpProxy.getPortOrDefault(8080)), config.isSecureProxy()));
+            Address proxyAddress = new Address(httpProxy.getHost(), httpProxy.getPortOrDefault(8080));
+            HttpProxy proxy = new HttpProxy(proxyAddress, config.isSecureProxy());
+            httpClient.getProxyConfiguration().addProxy(proxy);
+            if (config.getHttpProxyUser().isPresent() && config.getHttpProxyPassword().isPresent()) {
+                httpClient.getAuthenticationStore().addAuthentication(
+                        new BasicAuthentication(proxy.getURI(), ANY_REALM, config.getHttpProxyUser().get(), config.getHttpProxyPassword().get()));
```
Comment (Praveen2112) [inline http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java:311]: Can we apply them pre-emptively - So we could avoid the roundtrips ?

Comment (Praveen2112) [reply http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java:311]: httpClient.getAuthenticationStore().addAuthenticationResult(
                        new BasicAuthentication.BasicResult(
                                proxy.getURI(), 
                                HttpHeader.PROXY_AUTHORIZATION, 
                                config.getHttpProxyUser().get(), 
                                config.getHttpProxyPassword().get()));

File: http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java [hunk 2]
```diff
@@ -275,6 +277,30 @@ public HttpClientConfig setHttpProxy(HostAndPort httpProxy)
...
+        this.httpProxyUser = Optional.of(httpProxyUser);
+        return this;
+    }
+
+    public Optional<String> getHttpProxyPassword()
+    {
+        return httpProxyPassword;
+    }
+
+    @Config("http-client.http-proxy.password")
```
Comment (wendigo) [inline http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java:297]: @ConfigSecuritySensitive

File: http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java [hunk 3]
```diff
@@ -301,7 +303,13 @@ protected void configure(SelectableChannel selectable)
         }
         HostAndPort httpProxy = config.getHttpProxy();
         if (httpProxy != null) {
-            httpClient.getProxyConfiguration().addProxy(new HttpProxy(new Address(httpProxy.getHost(), httpProxy.getPortOrDefault(8080)), config.isSecureProxy()));
+            Address proxyAddress = new Address(httpProxy.getHost(), httpProxy.getPortOrDefault(8080));
+            HttpProxy proxy = new HttpProxy(proxyAddress, config.isSecureProxy());
+            httpClient.getProxyConfiguration().addProxy(proxy);
+            if (config.getHttpProxyUser().isPresent() && config.getHttpProxyPassword().isPresent()) {
+                httpClient.getAuthenticationStore().addAuthenticationResult(
+                        new BasicAuthentication.BasicResult(proxy.getURI(), HttpHeader.PROXY_AUTHORIZATION, config.getHttpProxyUser().get(), config.getHttpProxyPassword().get()));
```
Comment (wendigo) [inline http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java:311]: static import BasicResult

File: http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java [hunk 4]
```diff
@@ -1091,9 +1099,9 @@ public void dumpStdErr()
     public String dumpAllDestinations()
     {
         return format("%s\t%s\t%s\t%s\t%s\n", "URI", "queued", "request", "wait", "response") +
```
Comment (wendigo) [inline http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java:1101]: revert, unrelated

File: http-client/src/test/java/io/airlift/http/client/jetty/AbstractHttpClientTestHttpProxyAuthN.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            String authorization = req.getHeader(HttpHeader.PROXY_AUTHORIZATION.asString());
+            if (authorization == null) {
+                resp.setStatus(HttpStatus.PROXY_AUTHENTIATION_REQUIRED.code());
+                resp.setHeader(HttpHeader.PROXY_AUTHENTICATE.asString(), "Basic realm=\"Proxy Realm\"");
+            }
+            else {
+                String prefix = "Basic ";
+                if (authorization.startsWith(prefix)) {
+                    String attempt = authorization.substring(prefix.length());
+                    String credentials = Base64.getEncoder().encodeToString((this.user + ":" + this.password).getBytes(StandardCharsets.ISO_8859_1));
```
Comment (wendigo) [inline http-client/src/test/java/io/airlift/http/client/jetty/AbstractHttpClientTestHttpProxyAuthN.java:82]: you can keep the encoded string in the final field in the class

File: http-client/src/test/java/io/airlift/http/client/jetty/AbstractHttpClientTestHttpProxyAuthN.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+        @Override
+        protected void service(HttpServletRequest req, HttpServletResponse resp)
+                throws ServletException, IOException
+        {
+            String authorization = req.getHeader(HttpHeader.PROXY_AUTHORIZATION.asString());
+            if (authorization == null) {
+                resp.setStatus(HttpStatus.PROXY_AUTHENTIATION_REQUIRED.code());
+                resp.setHeader(HttpHeader.PROXY_AUTHENTICATE.asString(), "Basic realm=\"Proxy Realm\"");
```
Comment (wendigo) [inline http-client/src/test/java/io/airlift/http/client/jetty/AbstractHttpClientTestHttpProxyAuthN.java:76]: return;

and drop else {}

File: http-client/src/test/java/io/airlift/http/client/jetty/TestJettyHttpClientHttpProxyAuthN.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.http.client.jetty;
+
+public class TestJettyHttpClientHttpProxyAuthN
+        extends AbstractHttpClientTestHttpProxyAuthN
```
Comment (wendigo) [inline http-client/src/test/java/io/airlift/http/client/jetty/TestJettyHttpClientHttpProxyAuthN.java:17]: Do we need abstract class? Merge these classes together

Comment (marcinsbd) [reply http-client/src/test/java/io/airlift/http/client/jetty/TestJettyHttpClientHttpProxyAuthN.java:17]: If we merge them, the TestJettyHttpClientHttpProxyAuthN will need to have two constructor, one for init the class, and the second with parameter for the extension in derived class. However, the Junit framework demands single constructor for the test class.

**Review by wendigo — APPROVED** [review]

**Pull request merged**